RequestCache KeyError in SearchCommunity

The problem is in current `devel`:
```
ERROR   1423136054.27        dispersy:1724  exception during handle_callback for torrent-collect-response
Traceback (most recent call last):
  File "/home/lfei/workspace/tud/p2p/tribler/Tribler/dispersy/dispersy.py", line 1719, in _update
    messages[0].handle_callback(messages)
  File "/home/lfei/workspace/tud/p2p/tribler/Tribler/community/search/community.py", line 489, in on_torrent_collect_response
    self.request_cache.pop(u"ping", message.payload.identifier)
  File "/home/lfei/workspace/tud/p2p/tribler/Tribler/dispersy/requestcache.py", line 196, in pop
    cache = self._identifiers.pop(identifier)
KeyError: u'ping:25657'
ERROR   1423136054.29        dispersy:1724  exception during handle_callback for torrent-collect-response
Traceback (most recent call last):
  File "/home/lfei/workspace/tud/p2p/tribler/Tribler/dispersy/dispersy.py", line 1719, in _update
    messages[0].handle_callback(messages)
  File "/home/lfei/workspace/tud/p2p/tribler/Tribler/community/search/community.py", line 489, in on_torrent_collect_response
    self.request_cache.pop(u"ping", message.payload.identifier)
  File "/home/lfei/workspace/tud/p2p/tribler/Tribler/dispersy/requestcache.py", line 196, in pop
    cache = self._identifiers.pop(identifier)
KeyError: u'ping:41426'
```

Probably related to this error:

<pre>
 CRITICAL 1423226122.23     minitwisted:105   MINITWISTED CRASHED
 ERROR   1423226122.23     minitwisted:106   MINITWISTED CRASHED 
 Traceback (most recent call last):
   File "/home/rob/git/tribler/Tribler/Core/DecentralizedTracking/pymdht/core/minitwisted.py", line 103, in run2
    self.run_one_step()
  File "/home/rob/git/tribler/Tribler/Core/DecentralizedTracking/pymdht/core/minitwisted.py", line 179, in run_one_step
    datagram_received)
  File "/home/rob/git/tribler/Tribler/Core/DecentralizedTracking/pymdht/core/controller.py", line 262, in on_datagram_received
    callback_f(lookup_id, peers, msg.src_node)
  File "/home/rob/git/tribler/Tribler/community/tunnel/hidden_community.py", line 227, in dht_callback
    self.create_key_request(info_hash, peer)
  File "/home/rob/git/tribler/Tribler/community/tunnel/hidden_community.py", line 240, in create_key_request
    cache = self.request_cache.add(KeyRequestCache(self, circuit, sock_addr, info_hash))
  File "/home/rob/git/tribler/Tribler/community/tunnel/hidden_community.py", line 60, in __init__
    super(KeyRequestCache, self).__init__(community.request_cache, u"key-request")
  File "/home/rob/git/tribler/Tribler/dispersy/requestcache.py", line 50, in __init__
    number = RandomNumberCache.find_unclaimed_identifier(request_cache, prefix)
  File "/home/rob/git/tribler/Tribler/dispersy/requestcache.py", line 57, in find_unclaimed_identifier
    if not request_cache.has(prefix, number):
  File "/home/rob/git/tribler/Tribler/dispersy/requestcache.py", line 172, in has
    assert isInIOThread(), "RequestCache must be used on the reactor's thread"
AssertionError: RequestCache must be used on the reactor's thread
MINITWISTED CRASHED (see logs)</pre>

--- a/Tribler/community/search/community.py
+++ b/Tribler/community/search/community.py
@@ -488,7 +488,7 @@ class SearchCommunity(Community):
         self._create_pingpong(u"torrent-collect-response", candidates, identifiers)
         self._process_collect_request_response(messages)
 
-    def on_torrent_collect_response(self, messages, verifyRequest=True):
+    def on_torrent_collect_response(self, messages):
         self._process_collect_request_response(messages)
 
         for message in messages:

--- a/Tribler/community/search/community.py
+++ b/Tribler/community/search/community.py
@@ -469,7 +469,6 @@ class SearchCommunity(Community):
                     break
 
             if remove:
-                self._logger.debug(u"no response on ping, removing from taste_buddies %s", self.candidate)
                 self.community.taste_buddies.remove(remove)
 
     def create_torrent_collect_requests(self, candidates=None):
@@ -491,14 +490,18 @@ class SearchCommunity(Community):
     def on_torrent_collect_response(self, messages):
         self._process_collect_request_response(messages)
 
-        for message in messages:
-            self.request_cache.pop(u"ping", message.payload.identifier)
-
     def _process_collect_request_response(self, messages):
         to_insert_list = []
         to_collect_dict = {}
         to_popularity_dict = {}
         for message in messages:
+            # check if the identifier is still in the request_cache because it could be timed out
+            if not self.request_cache.has(u"ping", message.payload.identifier):
+                self._logger.warn(u"message from %s cannot be found in the request cache, skipping it",
+                                  message.candidate)
+                continue
+            self.request_cache.pop(u"ping", message.payload.identifier)
+
             if message.payload.hashtype == SWIFT_INFOHASHES:
                 for infohash, seeders, leechers, ago in message.payload.torrents:
                     if not infohash:

